### DISCUSSION TITLE
Validation of `@defer`/`@stream` label arguments

### DISCUSSION THREAD

**Discussion by robrichard** [graphql/defer-stream-wg#35] [Architectural Decision Records]
## Context

raised by @BoD https://github.com/graphql/graphql-spec/pull/742#discussion_r814635618

> Should there be a formal validation rule for the label unicity? For instance it could clarify whether labels should be unique in an operation + any fragments it references.

The `label` argument on `@defer` and `@stream` is used by GraphQL clients to identify the data from patch responses and associate it with the correct fragment. If provided, the GraphQL Server must add it to the payload.

Validating this argument runs into the same issue that was discussed in [@stream initialCount validation](https://github.com/robrichard/defer-stream-wg/discussions/24); the validation stage can happen before variables are known. 

Three options considered:

1. No validation of the label argument
2. Raise a field error during execution if duplicate labels are encountered
3. Validation that prevents variables from being used in the label argument and checks for uniqueness. I'm not sure if there's value in having dynamic labels 

## Decision
Discussed at the [March 2022 WG meeting](https://github.com/graphql/graphql-wg/blob/main/notes/2022/2022-03-03.md#defer--stream-discussion-30m-rob-richard) and decided to add validation preventing variables in labels, and a validation enforcing uniqueness. 

There haven't been any use cases raised that require variables for label values. This could be loosened later without breaking backwards compatibility if additional use cases are discovered.

## Status
- [x] defined in spec
- [x] implemented in graphql-js

**Comment by glasser** [comment]
3 sounds reasonable to me. If this ends up preventing real use cases, we can always allow it later.

**Comment by robrichard** [comment]
This was discussed at the [March 2022 WG meeting](https://github.com/graphql/graphql-wg/blob/main/notes/2022/2022-03-03.md#defer--stream-discussion-30m-rob-richard) and we decided to add validation preventing variables in labels, and a validation enforcing uniqueness. Spec and reference implementation PRs have both been updated to reflect this.

**Comment by robrichard** [comment]
from @yaacovcr

> Would it be possible to allow variables for label, as long as the same variable is not used more than once?

**Reply by robrichard** [reply]
I think that could work, it could be explored further if the current validation ends up preventing real use cases

**Reply by yaacovCR** [reply]
I guess you could run into an issue if different variables happened to have the same value, but I am not sure if we really need to validate for that, presumably if you are using variables as labels, you are doing so for a reason and ensuring they are unique on your own...

**Comment by BoD** [comment]
Sorry to come back to this topic again! A question arose while implementing this validation: is "document" meant to mean the whole document across multiple operations? For instance, this would mean the following is invalid:

```graphql
query Query1 {
  field {
    ...fieldDetails1 @defer(label: "details")
  }
}

query Query2 {
  field {
    ...fieldDetails2 @defer(label: "details")
  }
}
```

even though this can never cause any ambiguity?

**Reply by robrichard** [reply]
@BoD "document" is intended to mean the entire document, not a specific operation in a document that contains multiple operations. I think it makes the validation significantly simpler (at least in the reference implementation) if the labels must be unique across the entire document instead of each operation. I think we can consider loosening this restriction if it's blocking any use cases.

**Reply by BoD** [reply]
Thanks a lot for the clarification.  

In our case, implementing both versions is equivalent - we were just wondering if scoping the rule to the whole document was useful in a way that we missed.

**Reply by glasser** [reply]
@BoD Worth noting that in `graphql-js`, the "validation" stage does not actually take the specified `operationName` as an input.
